network_plugin/keypair.py

```diff
@@ -10,9 +20,70 @@ def creation_validation(**kwargs):
         check availability of path used in field private_key_path of
         node properties
     """
-    key = ctx.node.properties.get('private_key_path')
-    if key:
+    key = ctx.node.properties.get(PRIVATE_KEY)
+    if key.get(PATH):
         key_path = os.path.expanduser(key)
         if not os.path.isfile(key_path):
             raise cfy_exc.NonRecoverableError(
                 "Private key file {0} is absent".format(key_path))
+
+
+@operation
+def create(**kwargs):
+    ctx.instance.runtime_properties[PUBLIC_KEY] = {}
+    ctx.instance.runtime_properties[PRIVATE_KEY] = {}
+    ctx.instance.runtime_properties[PUBLIC_KEY][USER] = \
+        ctx.node.properties.get(PUBLIC_KEY).get(USER)
+    if ctx.node.properties.get(AUTO_GENERATE):
+        ctx.logger.info("Generating ssh keypair")
+        public, private = _generate_pair()
+        ctx.instance.runtime_properties[PRIVATE_KEY][PATH] = _create_path()
+        ctx.instance.runtime_properties[PRIVATE_KEY][KEY] = private
+        ctx.instance.runtime_properties[PUBLIC_KEY][KEY] = public
+        _save_key_file(ctx.instance.runtime_properties[PRIVATE_KEY][PATH],
+                       ctx.instance.runtime_properties[PRIVATE_KEY][KEY])
+    else:
+        ctx.instance.runtime_properties[PUBLIC_KEY][KEY] = \
+            ctx.node.properties.get(PUBLIC_KEY).get(KEY)
+        ctx.instance.runtime_properties[PRIVATE_KEY][KEY] = \
+            ctx.node.properties[PRIVATE_KEY].get(KEY)
+        ctx.instance.runtime_properties[PRIVATE_KEY][PATH] = \
+            ctx.node.properties[PRIVATE_KEY].get(PATH)
+        if ctx.node.properties[PRIVATE_KEY].get(KEY):
+            ctx.instance.runtime_properties[PRIVATE_KEY][PATH] = _create_path()
+            _save_key_file(ctx.instance.runtime_properties[PRIVATE_KEY][PATH],
+                           ctx.instance.runtime_properties[PRIVATE_KEY][KEY])
+
+
+@operation
+def delete(**kwargs):
+    if ctx.node.properties[AUTO_GENERATE]:
+        _delete_key_file(ctx.instance.runtime_properties[PRIVATE_KEY][PATH])
+    else:
+        if ctx.node.properties[PRIVATE_KEY].get(KEY):
+            _delete_key_file(ctx.instance.runtime_properties[PRIVATE_KEY][PATH])
+    del ctx.instance.runtime_properties[PRIVATE_KEY]
+    del ctx.instance.runtime_properties[PUBLIC_KEY]
+
+
+def _generate_pair():
+    Random.atfork()  # uses for strong key generation
+    key = RSA.generate(2048)
+    private_value = key.exportKey('PEM')
+    public_value = key.publickey().exportKey('OpenSSH')
+    return public_value, private_value
+
+
+def _create_path():
+    return '~/.ssh/{}_private.key'.format(ctx.instance.id)
```

Comment:
We need to take into account that the cloudify server is multi-tenant and although the instance.id may be unique, (I am not sure) for the blueprint, I am pretty sure it is ot unique across blueprints. I suggest storing it in the deployment virtenv folder instead of the shared ~/.ssh folder.